### ISSUE TITLE
Investigate rendering backends, such as gfx-rs or wgpu-rs or glow

### ISSUE THREAD

**Issue by iceiix** [iceiix/stevenarella#34]
Currently, steven uses OpenGL 3.2 core profile, through ~SDL2 (#35)~ and [gl_generator](https://crates.io/crates/gl_generator) + [khronos_api](https://crates.io/crates/khronos_api) with a custom but lightweight wrapper, steven_gl: https://github.com/iceiix/steven/tree/updates/gl + https://github.com/iceiix/steven/blob/updates/src/gl/mod.rs. Is there a better way to do this? (3D graphics)

May help with graphical glitches such as https://github.com/iceiix/steven/issues/25, and/or cross-platform compatibility

**Comment by iceiix** [comment]
gfx-rs is coming up again in #92 WebAssembly support, once it supports wasm32-unknown-unknown/WebGL target, it will be quite appealing for this project. grovesNL is working on adding that, using his "GL on Whatever" set of bindings: https://github.com/grovesNL/glow to target OpenGL + OpenGL ES + WebGL avoiding target-specific code.

As an intermediate step, may want to see if can replace the homegrown steven_gl wrapper with glow? Though ultimately, gfx-rs may be the way to go. Start porting to it now? (for native at first)

**Comment by iceiix** [comment]
https://gfx-rs.github.io/2019/03/06/wgpu.html https://www.reddit.com/r/rust/comments/ay3olj/the_rise_of_wgpurs/

**Comment by iceiix** [comment]
https://wiki.alopex.li/AGuideToRustGraphicsLibraries2019
https://old.reddit.com/r/rust/comments/bjnxd3/a_guide_to_rust_graphics_libraries_as_of_2019/
https://old.reddit.com/r/rust_gamedev/comments/bjsk1q/a_guide_to_rust_graphics_libraries_as_of_2019/

No clear winner still, but there are some important recent developments:

2019-04-26 https://phaazon.net/blog/pre-luminance-n-random-thoughts - for 1.0.0 the author is planning "At least one WebGL implementation and it must be WASM compatible."

2019-04-18 https://amethyst.rs/blog/moss-grant-announce/ - Amethyst was awarded a $10,000 grant for making it run in the browser with WebAssembly, and they plan to help where possible with:

https://github.com/rust-windowing/winit/pull//797 stdweb support for eventloop 2.0 
https://github.com/gfx-rs/gfx/pull//2554 [WIP] Add wasm32-unknown-unknown/WebGL support

WebGPU working group https://www.w3.org/community/gpu/ https://github.com/gpuweb/ (not to be confused with https://github.com/webgpu and https://webgpu.com) is still active, as is https://github.com/gfx-rs/wgpu/, but isn't shipping in browsers yet. Targeting [wgpu](https://crates.io/crates/wgpu) may very well be a good option, and the wgpu developers agree:

https://old.reddit.com/r/rust/comments/bjnxd3/a_guide_to_rust_graphics_libraries_as_of_2019/emeznnf/

> My vision is that 95% of applications (that want to program graphics workloads directly instead of relying on engines like AmethysT) would be fine at building on top of `wgpu` even if they don't care about the Web. It's just a nice abstraction level to work with, and the safety guarantee plays quite nicely with Rust.

**Comment by iceiix** [comment]
Posted more thoughts at https://www.reddit.com/r/rust/comments/bjnxd3/a_guide_to_rust_graphics_libraries_as_of_2019/emfekoe/ (in https://www.reddit.com/r/rust/comments/bjnxd3/a_guide_to_rust_graphics_libraries_as_of_2019/), but looking at https://github.com/gfx-rs/wgpu/ more closely, it already supports Vulkan (Windows/Linux), D3D12/11 (Windows), and Metal (macOS/iOS) platforms. Could I port to `wgpu` to not regress on any supported platforms, while opening the door for WebGPU support when available?

**Comment by iceiix** [comment]
https://users.rust-lang.org/t/help-porting-a-large-complex-application-to-the-web-wasm32-unknown-emscripten-wasm32-unknown-unknown-wasm32-wasi/28527/5

https://github.com/iceiix/stevenarella/issues/171

**Comment by iceiix** [comment]
https://old.reddit.com/r/rust_gamedev/comments/brehgw/rendy_02_released/ has some more insights:

> You're right, basically users can write against wgpu-rs to target both native and web: 
> - when they target native, wgpu-native (the native WebGPU implementation on top of gfx-hal) is used and it runs on Vulkan/Metal/DX12/etc. 
> - when they target wasm32-unknown-unknown, the wgpu-native layer disappears and the browser's API is used instead

warming up to the idea of targeting wgpu-rs...

**Comment by iceiix** [comment]
https://github.com/Gordon-F/crossplatform_gfx_hal_example

https://www.reddit.com/r/rust_gamedev/comments/bv7w2f/wgpurs_now_uses_rendy_to_manage_memory_and/

Looks like wgpu-rs is again a good choice, as for "4. wgpu-rs for the lowest safe level",

**Comment by iceiix** [comment]
> grovesNL is working on adding that, using his "GL on Whatever" set of bindings: https://github.com/grovesNL/glow to target OpenGL + OpenGL ES + WebGL avoiding target-specific code.
> 
> As an intermediate step, may want to see if can replace the homegrown steven_gl wrapper with glow? Though ultimately, gfx-rs may be the way to go.

https://github.com/rust-windowing/glutin now recommends glow:

> Emscripten with asmjs
>
> Emscripten support has been deprecated in favor of platforms like stdweb. To get an OpenGL > context on these platforms, please use crates like [glow](https://crates.io/crates/glow) instead.

https://github.com/grovesNL/glow

> GL on Whatever: a set of bindings to run GL anywhere (Open GL, OpenGL ES, and WebGL) and avoid target-specific code.

Looks great. The [example](https://github.com/grovesNL/glow/tree/master/examples/hello) shows targeting native through glutin, native through sdl2, and the web. Is this the cross-compatible crate I have been searching for?  How hard would it be to adapt to use glow instead of glutin?

**Comment by iceiix** [comment]
https://github.com/sebcrozet/kiss3d http://kiss3d.org

**Comment by iceiix** [comment]
Update: gfx-hal 0.3.0 is released, continues to march forward: https://www.reddit.com/r/rust/comments/cnvfrm/gfxhal03_release_notes/

Most notably, they added wasm32-unknown-unknown support to the GL backend so it [runs on the web](https://github.com/gfx-rs/gfx/pull/2554#issuecomment-474206720), and switched from gl-rs to [glow](https://github.com/grovesNL/glow). 

Looking more and more that *glow* is the way to go. Not to rule out gfx-rs but would be a larger change. Ought to replace gl-generator/khronos_api (from steven_gl) with glow then see where we're at with #171 wasm support.

**Comment by iceiix** [comment]
icefoxen (no relation) has posted an excellent (as usual) "State of GGEZ 2020" report with many points also relevant to this project: https://wiki.alopex.li/TheStateOfGGEZ2020

A few takeaways: gfx-hal and rendy are promising cross-API compatible abstraction layers, a Vulkan-like API, but 

> I still can’t use something better than OpenGL on the web. My gut response to that is to say 'then help make it happen', but I’m just too tired to yak-shave that hard. So, for now, I’m giving up.

What about WebGPU? Not standardized yet, probably will happen some day with wgpu using gfx-hal and rendy so "its ability to run on Web is still tied to theirs". 

The author speaks positively of newer versions of OpenGL: 

> WebGL 2 is now an explicit, strict subset of OpenGL ES 3.0, which is now an *explicit, strict* subset of OpenGL 4.3. So if you write your code and shaders to run on WebGL 2, they can run anywhere that supports ES 3 or OpenGL 4.3, without modifications, for realsies, no takebacks. (At least before the inevitable driver bugs occur, that is.) 

This project currently targets OpenGL 3.2 and OpenGL ES 2.0, src/main.rs:

```rust
        .with_gl(glutin::GlRequest::GlThenGles{opengl_version: (3, 2), opengles_version: (2, 0)})
```

should update to OpenGL 4.3 and OpenGL ES 3.0, to target WebGL 2?

The author of ggez came to the same conclusion I was leaning towards:

> Conclusions: So yeah, I’ve been rewriting `ggez`’s graphics backend in pure OpenGL, using the excellent glow `crate` to provide a modicum of abstraction between normal OpenGL and WebGL. It actually works pretty well

glow it is! (At least until gfx or wgpu are more ready)

**Comment by iceiix** [comment]
More progress on wgpu-rs: https://gfx-rs.github.io/2020/04/21/wgpu-web.html wgpu-rs on the web

> gfx-rs is a Rust project aiming to make low-level GPU programming portable with low overhead. It’s a single Vulkan-like Rust API with multiple backends that implement it: Direct3D 12/11, Metal, Vulkan, and even OpenGL.
>
> wgpu-rs is a Rust project on top of gfx-rs that provides safety, accessibility, and even stronger portability.

**Comment by iceiix** [comment]
Decided for now to go with [glow](https://github.com/grovesNL/glow), GL on Whatever. A lightweight wrapper, supporting up to OpenGL 4.6 on native; we currently target OpenGL 3.2 on native so this works. For web, we could target WebGL 2 through glow (and also possibly OpenGL ES 3.0 for native).

Merged glow support in #262, only working with OpenGL at the moment but see #446 and #451 for progress on getting WebGL working.
